Backports 0.17 pr19 (#3741)
dashpay/dash: doc/developer-notes.md

Developer Notes
===============

<!-- markdown-toc start -->
**Table of Contents**

- [Developer Notes](#developer-notes)
    - [Coding Style](#coding-style)
    - [Doxygen comments](#doxygen-comments)
    - [Development tips and tricks](#development-tips-and-tricks)
        - [Compiling for debugging](#compiling-for-debugging)
        - [Compiling for gprof profiling](#compiling-for-gprof-profiling)
        - [debug.log](#debuglog)
        - [Testnet and Regtest modes](#testnet-and-regtest-modes)
        - [DEBUG_LOCKORDER](#debug_lockorder)
        - [Valgrind suppressions file](#valgrind-suppressions-file)
        - [Compiling for test coverage](#compiling-for-test-coverage)
    - [Locking/mutex usage notes](#lockingmutex-usage-notes)
    - [Threads](#threads)
    - [Ignoring IDE/editor files](#ignoring-ideeditor-files)
- [Development guidelines](#development-guidelines)

>     - [General Bitcoin Core](#general-bitcoin-core)

Reviewer: ```suggestion
    - [General Dash Core](#general-dash-core)
```

# A coverage report will now be accessible at `./test_dash.coverage/index.html`.
```

**Sanitizers**

> Bitcoin can be compiled with various "sanitizers" enabled, which add

Reviewer: ```suggestion
Dash Core can be compiled with various "sanitizers" enabled, which add
```

- src/univalue
  - Upstream at https://github.com/jgarzik/univalue ; report important PRs to Core to avoid delay.

Upgrading LevelDB
---------------------

Extra care must be taken when upgrading LevelDB. This section explains issues
you must be aware of.

### File Descriptor Counts

In most configurations we use the default LevelDB value for `max_open_files`,
which is 1000 at the time of this writing. If LevelDB actually uses this many
file descriptors it will cause problems with Bitcoin's `select()` loop, because
it may cause new sockets to be created where the fd value is >= 1024. For this
reason, on 64-bit Unix systems we rely on an internal LevelDB optimization that
uses `mmap()` + `close()` to open table files without actually retaining
references to the table file descriptors. If you are upgrading LevelDB, you must
sanity check the changes to make sure that this assumption remains valid.

In addition to reviewing the upstream changes in `env_posix.cc`, you can use `lsof` to
check this. For example, on Linux this command will show open `.ldb` file counts:

```bash

> $ lsof -p $(pidof bitcoind) |\

Reviewer: ```suggestion
$ lsof -p $(pidof dashd) |\
```
